dpctl/dparray.py

```diff
@@ -0,0 +1,257 @@
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+        else:
+            dprint("dparray::ndarray __new__ buffer not None and not sycl_usm")
+            nelems = np.prod(shape)
+            # must copy
+            ar = np.ndarray(
+                shape,
+                dtype=dtype,
+                buffer=buffer,
+                offset=offset,
+                strides=strides,
+                order=order,
+            )
+            nbytes = int(ar.nbytes)
+            buf = MemoryUSMShared(nbytes)
+            new_obj = np.ndarray.__new__(
+                subtype,
+                shape,
+                dtype=dtype,
+                buffer=buf,
+                offset=0,
+                strides=strides,
+                order=order,
+            )
+            np.copyto(new_obj, ar, casting="no")
+            if hasattr(new_obj, array_interface_property):
+                dprint("buffer None new_obj already has sycl_usm")
+            else:
+                dprint("buffer None new_obj will add sycl_usm")
+                setattr(new_obj, array_interface_property, {})
+            return new_obj
+
+    def __array_finalize__(self, obj):
+        dprint("__array_finalize__:", obj, hex(id(obj)), type(obj))
+        # When called from the explicit constructor, obj is None
+        if obj is None:
+            return
+        # When called in new-from-template, `obj` is another instance of our own
+        # subclass, that we might use to update the new `self` instance.
+        # However, when called from view casting, `obj` can be an instance of any
+        # subclass of ndarray, including our own.
+        if hasattr(obj, array_interface_property):
+            return
+        if isinstance(obj, numba.core.runtime._nrt_python._MemInfo):
+            mobj = obj
+            while isinstance(mobj, numba.core.runtime._nrt_python._MemInfo):
+                dprint("array_finalize got Numba MemInfo")
+                ea = mobj.external_allocator
+                d = mobj.data
+                dprint("external_allocator:", hex(ea), type(ea))
+                dprint("data:", hex(d), type(d))
+                dppl_rt_allocator = numba.dppl._dppl_rt.get_external_allocator()
+                dprint(
+                    "dppl external_allocator:",
+                    hex(dppl_rt_allocator),
+                    type(dppl_rt_allocator),
+                )
+                dprint(dir(mobj))
+                if ea == dppl_rt_allocator:
+                    return
+                mobj = mobj.parent
+                if isinstance(mobj, ndarray):
+                    mobj = mobj.base
+        if isinstance(obj, np.ndarray):
+            ob = self
+            while isinstance(ob, np.ndarray):
+                if hasattr(obj, array_interface_property):
```

Comment:
This needs to be `hasattr(ob, array_interface_property)`.

Reply:
Done.